6347 Cache the number of each component type assigned to devices/VMs

arthanson opened the pull request:
### Fixes: #6347 

Adds counter fields to cache related model counts.  Does this by adding a CachedCounterField type that hooks up post-save and post-delete signals on the related model, these hooked up by calling connect_counter in apps.py

also has change tracking, there is a TrackingModelMixin which I added to the base classes - by default it does nothing but if you set change_tracking_fields on the model to the fields to track it will do change tracking on them.  The connect_counter method automatically does the correct fields (generally device_id) and then checks this on post_save signal  to see if it has been moved and updates the counts on the old and new parent models appropriately.

The change tracking is probably useful elsewhere - you can check model.tracker.changed to see any changed fields.  Note: these fields must be listed in change_tracking_fields.

arthanson committed:
6347 initial counter field

arthanson committed:
6347 connect the counters

arthanson committed:
6347 migrate initial counts

arthanson committed:
6347 migrate initial counts

arthanson committed:
6347 fix count

arthanson committed:
6347 comment

arthanson committed:
6347 reindent migration

arthanson committed:
6347 virtual machine interface count

arthanson committed:
6347 add missing migration

arthanson committed:
6347 only update count if created

arthanson committed:
6347 clarify name

kkthxbye-code commented:
I know this is a draft, so apologies if this is premature. Looking at the code it doesn't look like it currently take into account that InventoryItems can be moved between devices.

It might also be an idea to add a management command to re-calculate counts, if they ever get out of sync.

arthanson committed:
6347 track item move

arthanson committed:
6347 add management command

arthanson commented:
@kkthxbye-code could definitely use a more thorough review on this one, any feedback welcome.

jeremystretch commented on netbox/utilities/counter.py:
```diff
@@ -0,0 +1,91 @@
+            raise TypeError(
+                f"{self.counter_name} should be a CounterCacheField on {self.parent_model}, but is {type(counter_field)}"
+            )
+        if not isinstance(parent_model, TrackingModelMixin):
```
```suggestion
        if not isinstance(self.parent_model, TrackingModelMixin):
```

jeremystretch commented on netbox/utilities/mixins.py:
```diff
@@ -0,0 +1,54 @@
+
+
+class TrackingModelMixin(object):
+    change_tracking_fields = []
```
Because this is declared as a class attribute, the list is being shared across _all_ models:

```python
>>> ConsolePort.change_tracking_fields
['device_id', 'virtual_machine_id']
```

Instead of defining this on the model, could we create a registry entry for mapping models to their tracked fields? This would probably be a bit cleaner.

arthanson replied:
made an instance var

jeremystretch commented on netbox/utilities/fields.py:
```diff
@@ -143,3 +143,12 @@ def gfk_key(obj):
         )
+
+
+class CounterCacheField(models.BigIntegerField):
```
Need to add this to `__all__`

arthanson committed:
Merge branch 'feature' into 6347-cache-counts

arthanson committed:
6347 update ref to parent_model, migration

arthanson committed:
6347 make change_tracking_fields instance var

arthanson committed:
6347 set initial counts for Device in migration

jeremystretch committed:
Merge branch 'feature' into 6347-cache-counts

jeremystretch commented on netbox/utilities/mixins.py:
```diff
@@ -0,0 +1,54 @@
+
+    @property
+    def tracker(self):
+        if hasattr(self._state, "_tracker"):
+            tracker = self._state._tracker
+        else:
+            tracker = self._state._tracker = Tracker(self)
+        return tracker
```
```suggestion
        if not hasattr(self._state, "_tracker"):
            self._state._tracker = Tracker(self)
        return self._state._tracker
```

jeremystretch commented on netbox/utilities/mixins.py:
```diff
@@ -0,0 +1,54 @@
+class Tracker(object):
+    def __init__(self, instance):
+        self.instance = instance
+        self.newly_created = False
```
Do we need the `newly_created` attribute? It doesn't appear to ever be read.

jeremystretch commented on netbox/utilities/mixins.py:
```diff
@@ -0,0 +1,54 @@
+        self.instance = instance
+        self.newly_created = False
+        self.changed = {}
+        self.tracked_fields = self.instance.change_tracking_fields
```
This seems redundant. Won't the value of a Tracker instance's `tracked_fields` attribute always be equal to the value of `change_tracking_fields` on its instance?

jeremystretch commented on netbox/utilities/management/commands/calculate_cached_counts.py:
```diff
@@ -0,0 +1,36 @@
+from virtualization.models import VirtualMachine
+
+
+def recalculate_device_counts():
```
Ideally we'd have some mechanism to discover all the necessary fields, rather than defining them explicitly. Again, this is where utilizing the registry can help.

jeremystretch reviewed (changes requested):
Excellent work! Just working on simplifying and cleaning things up a bit at this point. Also, could you add some simple tests for the counters? A single relationship (e.g. device interfaces) should be sufficient.

arthanson committed:
Merge branch 'feature' into 6347-cache-counts

arthanson committed:
6347 review changes

arthanson committed:
6347 add tests

arthanson committed:
6347 make signal handlers use partial funcs

arthanson committed:
6347 optimize migrations

arthanson committed:
6347 optimize migrations

arthanson committed:
6347 optimize migrations

arthanson committed:
6347 move trackingmixin to specific models

arthanson committed:
6347 remove superfluous validate

arthanson committed:
6347 remove counters dict

arthanson committed:
6347 make connect_counter auto-discoverable

arthanson committed:
6347 make connect_counter auto-discoverable

arthanson committed:
6347 change_tracking_fields to class

arthanson commented:
@jeremystretch I added auto-discovery and cleaned up how you register / connect the fields, now you just supply it as params to the field:
```
    _interface_count = CounterCacheField(to_model='virtualization.VMInterface', to_field='virtual_machine')
```
and in apps.py ready function call:
```
        connect_counters(self)
```
I think this solves your issues with auto-discovery and the app registry?  Please let me know if any additional changes.

It would be nice if there was a global app init, then could just run through all models and wouldn't have to remember to put connect_counters in each application apps.py.

Should I document using this in the models doc?  Sort of an internal thing so not sure if this should go into the docs, but does effect people if they want to use this and add a field that should be cached.

arthanson committed:
6347 fix tracking mixin to tracked-classes

arthanson committed:
6347 fix tracking mixin to tracked-classes

jeremystretch committed:
Move change_tracking_fields from TrackingModelMixin to the registry

jeremystretch committed:
Remove _initialized flag from TrackingModelMixin

jeremystretch committed:
Simplify tests

jeremystretch committed:
Enable dynamic resolution of counter field mappings for the management command

jeremystretch committed:
Revert "Remove _initialized flag from TrackingModelMixin"

This reverts commit e7481f3bccf2e4918daedae114a4f5868f5b5504.

jeremystretch committed:
Create API for Tracker; add comments

jeremystretch committed:
Refactor counter logic to leverage the registry

jeremystretch committed:
Add test for moving a child model to a new parent

jeremystretch committed:
Rename counter fields for accessibility

jeremystretch committed:
Add component counts to API serializers

jeremystretch committed:
Add component count columns for devices & VMs

jeremystretch committed:
Add missing counter for module bays

jeremystretch committed:
View tabs should reference counter fields

jeremystretch committed:
Rename tracking utilities module

jeremystretch committed:
Fix tests
